### PR TITLE
✨ test framework: Add kubetest, kubernetesversions, log and ginkgo extensions

### PR DESCRIPTION
**Pull request by randomvariable**
Breaking out of #3593 

Mostly upstreaming from CAPA.

The kubetest package allows a consumer to run kubetest (i.e. Kubernetes conformance) against
given clusters.

The kubernetesversions package provides helpers to determine the latest CI release of Kubernetes
and adds debian/ubuntu injection script kustomizations for  based on prior work by @dims.

ginkgoextensions are common helpers that were added to CAPA, and are also present in the test/e2e
package.

Additionally add helpers to log running commands as well as gather JUnit files in a way that
is compatible with Prow.

Future PRs will add consumption of this in an e2e test and make CAPA consume these.

APIDiff will fail this PR, because a `GetRESTConfig` function has been added to the cluster proxy interface.  This is used so that a client-go discovery client can be used to discover the API server's kubernetes version. Believe this has minimal impact due to no third-party implementations of the interface.

Signed-off-by: Naadir Jeewa <jeewan@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #3569
Part of #2826

### CODE REVIEW

File: test/framework/kubernetesversions/zz_generated.bindata.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+	data, err := Asset(name)
+	if err != nil {
+		return err
+	}
+	info, err := AssetInfo(name)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(_filePath(dir, filepath.Dir(name)), os.FileMode(0755))
+	if err != nil {
```
Comment (vincepri) [inline test/framework/kubernetesversions/zz_generated.bindata.go:251]: if := ...; err != nil

File: test/framework/kubernetesversions/zz_generated.bindata.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		return err
+	}
+	err = ioutil.WriteFile(_filePath(dir, name), data, info.Mode())
+	if err != nil {
```
Comment (vincepri) [inline test/framework/kubernetesversions/zz_generated.bindata.go:255]: if := ...; err != nil

File: test/framework/kubernetesversions/zz_generated.bindata.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+		return err
+	}
+	err = os.Chtimes(_filePath(dir, name), info.ModTime(), info.ModTime())
+	if err != nil {
```
Comment (vincepri) [inline test/framework/kubernetesversions/zz_generated.bindata.go:259]: if := ...; err != nil

**Review by vincepri — COMMENTED** [review]
I can take a deeper look later

/assign @sedefsavas 
for another set of 👀

File: test/framework/kubernetesversions/zz_generated.bindata.go (hunk 1 above)
Comment (randomvariable) [reply test/framework/kubernetesversions/zz_generated.bindata.go:251]: generated file, nothing to do with me.

Comment (vincepri) [reply test/framework/kubernetesversions/zz_generated.bindata.go:251]: Whoops, missed it, no worries

File: test/framework/kubernetesversions/versions.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+const (
+	ciVersionURL     = "https://dl.k8s.io/ci/latest.txt"
+	stableVersionURL = "https://storage.googleapis.com/kubernetes-release/release/stable-%d.%d.txt"
+	tagPrefix        = "v"
+)
+
+// LatestCIRelease fetches the latest main branch Kubernetes version
+func LatestCIRelease() (string, error) {
```
Comment (sedefsavas) [inline test/framework/kubernetesversions/versions.go:35]: How about adding LatestRCRelease or latest release that is not an alpha release? I am thinking there is not much of a point in testing an alpha release. WDYT?

File: test/framework/kubetest/setup.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(b)), nil
+}
+
+func copy(src, dest string) error {
```
Comment (sedefsavas) [inline test/framework/kubetest/setup.go:46]:
```suggestion
func copyFile(srcFilePath, destFilePath string) error {
```

File: test/framework/kubernetesversions/bindata.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetesversions
+
+//go:generate sh -c "go-bindata -nometadata -pkg kubernetesversions -o zz_generated.bindata.go.tmp data && cat ../../../hack/boilerplate/boilerplate.generatego.txt zz_generated.bindata.go.tmp > zz_generated.bindata.go && rm zz_generated.bindata.go.tmp"
```
Comment (sedefsavas) [inline test/framework/kubernetesversions/bindata.go:19]: Do we need to add `go generate` to Makefile to generate this to update generated code if changes happen in the debian script for example?

File: test/framework/kubetest/run.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+	// ArtifactsDirectory is where conformance suite output will go
+	ArtifactsDirectory string
+	// Path to the kubetest e2e config file
+	ConfigFilePath string
+	// GinkgoNodes is the number of Ginkgo nodes to use
+	GinkgoNodes int
+	// GinkgoSlowSpecThreshold is time in s before spec is marked as slow
+	GinkgoSlowSpecThreshold int
+	// KubernetesVersion is the version of Kubernetes to test
+	KubernetesVersion string
```
Comment (sedefsavas) [inline test/framework/kubetest/run.go:63]: The version here should match with the cluster's Kubernetes version, right? Do we want to download a conformance image that matches with the cluster k8s version or conformance image version can be different than the cluster's k8s version?

File: test/framework/kubernetesversions/versions.go (hunk 4 above)
Comment (randomvariable) [reply test/framework/kubernetesversions/versions.go:35]: The CI release testing is used for CAPI providers which are Kubernetes release informing, such as CAPA.

File: test/framework/kubernetesversions/bindata.go (hunk 6 above)
Comment (randomvariable) [reply test/framework/kubernetesversions/bindata.go:19]: Added

File: test/framework/kubernetesversions/template.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+)
+
+type GenerateCIArtifactsInjectedTemplateForDebianInput struct {
+	ArtifactsDirectory        string
+	SourceTemplate            []byte
+	PlatformKustomization     []byte
+	KubeadmConfigTemplateName string
+	KubeadmControlPlaneName   string
+}
+
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply test/framework/kubernetesversions/template.go:50]: Done

File: test/framework/kubetest/run.go (hunk 7 above)
Comment (randomvariable) [reply test/framework/kubetest/run.go:64]: Added a server version discovery function which is consumed in Run so it's now optional. There may be a reason to run a later conformance suite on an older version, so have left the field in.

**Comment by benmoss** [comment]
/test pull-cluster-api-verify-external-links

just want to see if this works :smile:

File: test/framework/kubernetesversions/template.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	// PlatformKustomization is an SMP (strategic-merge-style) patch for adding
+	// platform specific kustomizations required for use with CI, such as
+	// referencing a specific image
+	PlatformKustomization []byte
+	// KubeadmConfigTemplateName is the name of the KubeadmConfigTemplate resource
+	// that needs to have the Debian install script injected. Defaults to "${CLUSTER_NAME}-md-0".
+	KubeadmConfigTemplateName string
+	// KubeadmControlPlaneName is the name of the KubeadmControlPlane resource
+	// that needs to have the Debian install script injected. Defaults to "${CLUSTER_NAME}-control-plane".
+	KubeadmControlPlaneName string
```
Comment (fabriziopandini) [inline test/framework/kubernetesversions/template.go:48]: Should we add Machine pool as well?

File: test/framework/log/log.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// log package implements an interface similar to klogr, and reuses much
+// of the same code, but makes the output appropriate for Ginkgo
+package log
```
Comment (fabriziopandini) [inline test/framework/log/log.go:19]: In the test framework there is already a custom logger under `test/framework/clusterctl/logger`. is it possible to make these two converge and get only one?

**Review by fabriziopandini — COMMENTED** [review]
@randomvariable thanks for this PR and sorry for lagging behind

This PR looks much more cleaner than the original one, and I'm looking forward to see this used by different providers
My only concern is about using this in docker, because in my experience there could be potential resource bottlenecks when executing the install scripts on several containers running on the same machines, and this can lead to flakes. But let's get this merged so we can test this in practice.

File: test/framework/kubernetesversions/template.go (hunk 9 above)
Comment (randomvariable) [reply test/framework/kubernetesversions/template.go:48]: We're only customising the PreKubeadmCommands and Files bits, so we shouldn't need to touch MachinePools do we?

**Comment by randomvariable** [comment]
> My only concern is about using this in docker, because in my experience there could be potential resource bottlenecks when executing the install scripts on several containers running on the same machines

Same, the other way would be Sonobuoy, but unfortunately we can't simultaneously import Sonobuoy and clusterctl framework because Sonobuoy imports clashing versions of API machinery. We can probably be revisit when we bump controller runtime in v1alpha4.

File: test/framework/log/log.go (hunk 10 above)
Comment (randomvariable) [reply test/framework/log/log.go:19]: Taking a look now.

File: test/framework/kubernetesversions/template.go (hunk 9 above)
Comment (fabriziopandini) [reply test/framework/kubernetesversions/template.go:48]: if this is the case, why do we need KubeadmConfigTemplateName and KubeadmControlPlaneName?

File: test/framework/log/log.go (hunk 10 above)
Comment (randomvariable) [reply test/framework/log/log.go]: I've removed it from this PR, will unify them in a separate PR

File: test/framework/kubernetesversions/template.go (hunk 9 above)
Comment (randomvariable) [reply test/framework/kubernetesversions/template.go:48]: those are the resources that get customised.

Comment (randomvariable) [reply test/framework/kubernetesversions/template.go:48]: Please look at https://github.com/kubernetes-sigs/cluster-api/pull/3652/files#diff-646b9b4cde9720e1732b0305f6f337edR174-R190

**Comment by randomvariable** [comment]
Any update on approval?

File: test/framework/kubernetesversions/template.go (hunk 9 above)
Comment (randomvariable) [reply test/framework/kubernetesversions/template.go:48]: Ah, the KubeadmConfig for a MachinePool. Makes sense. Will PR.

Comment (randomvariable) [reply test/framework/kubernetesversions/template.go:48]: https://github.com/kubernetes-sigs/cluster-api/pull/3730

**Pull request merged**